### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by EgorIlinskiy**

### CODE REVIEW

File: src/components/translation/index.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+const Translation:FC = (props)=>{
+
+    const local = navigator.language;
+    let language = useTypedSelector(selectLanguage);
+    const {SetEnglishLang, SetRussianLang} = useActions()
+
+    let text = English;
+   if(language === 'English') text = English
+    else if(language === 'Russian') text = Russian
```
Comment (IlyaRucavitcyn) [inline src/components/translation/index.tsx:20]: Think for a second - if this approach is scalable enough? What will happen if we have 5-6-7 languages? Maybe we need another approach?

File: src/store/actions/translation.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import {LanguageAction, LanguageActionTypes} from "../../types/languageTypes";
+
+export function SetEnglishLang():LanguageAction{
```
Comment (IlyaRucavitcyn) [inline src/store/actions/translation.ts:3]: Is there any reasons you use capital letters for function names?

File: src/types/cardTypes.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    name:string;
+    placeholder:string;
+    touched?:boolean;
+    value:string;
+    width?:string;
+}
+
+export interface ISelectInput{
+    id:string;
+    handleChange:any;
```
Comment (IlyaRucavitcyn) [inline src/types/cardTypes.ts:16]: Can we try to avoid `any`?

File: src/types/languageTypes.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+export interface LanguageTypes{
+    language:string;
+}
+
```
Comment (IlyaRucavitcyn) [inline src/types/languageTypes.ts:4]: Note: I saw a lot of places in the code where you have different number of spaces between functions or code snippets

**Review by IlyaRucavitcyn — CHANGES_REQUESTED** [review]

File: src/components/card/InputForms.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    const {AddNewCard, SwitchMode} = useActions()
+   const preSelected = useTypedSelector(state => state.card.preSelectedId)
+    const cardData = useTypedSelector(state => state.card).cards.filter(card=>{
+        return card.cardId === preSelected
+    })[0]
+
+
+    const initialValues:IFormValues = {
+        cardNumber: cardData !== undefined ? cardData.cardNumber : '',
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/InputForms.tsx]: Non actual file

File: src/components/card/InputForms.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+    const {values,
+        handleSubmit,
+        handleChange,
+        handleBlur,
+        errors,
+        touched,
+        setFieldValue
+    }:FormikProps<IFormValues> = useFormik({
+        initialValues:initialValues,
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/InputForms.tsx]: Refactor it

File: src/components/card/InputForms.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+        onSubmit:(values) => {
+            console.log(values)
+            if(values.saveCard) AddNewCard(values)
+    },
+        validationSchema:yup.object({
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/InputForms.tsx]: Moved schema to separate file

File: src/components/card/InputForms.tsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                                 placeholder={''}
+                                 errors={errors.cardCVV}
+                                 value={values.cardCVV}
+                                 label={'CVV / CVC'}
+                                 width={'auto'}
+                    />
+                    <span className={'payments-card-input-cvv-text'}>3 or 4 digits code</span>
+                </div>
+                <button type='submit' className='payments-card-submit'// disabled={Object.keys(errors).length>0}
+                 >Pay now</button>
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/InputForms.tsx]: Non actual files

File: src/components/card/InputForms.tsx [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            <div className={'payments-card-line'}></div>
+            <div className={'payments-card-settings'}>
+                <div className={'payments-card-settings-select'}
+                    onClick={()=>{SwitchMode()}}
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/InputForms.tsx]: There is no reason. Changed func names

File: src/components/card/SavedCards.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    const cards = useTypedSelector(state=>state.card.cards)
+
+    const {SetPreselectedCard} = useActions()
+
+
+
+
+    const renderCards = () =>{
+        return cards.map((card)=>{
+            console.log(card.cardId)
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/SavedCards.tsx]: Non actual file

File: src/components/card/options/optionsForSelect.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+export const yearOptions = generateYears()
+export const monthOptions = generateMonths();
+
+function generateYears():object[]{
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/options/optionsForSelect.ts:4]: This two functions need to generate options for select components. yearOptions generate object with 10+ years options, month generate object with 12 month

File: src/store/actions/card.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import {CardActions, CardActionTypes, IFormValues} from "../../types/cardTypes";
+
+export function AddNewCard(data:IFormValues):CardActions{
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/store/actions/card.ts]: Non actual file. Refactor also this.

File: src/store/reducers/cardReducers.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
+import {CardActions, CardActionTypes, CardStateType} from "../../types/cardTypes";
+import _ from "lodash";
+
+const initialState:CardStateType = {
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/store/reducers/cardReducers.ts]: Even if we need some initial state?

File: src/store/reducers/cardReducers.ts [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+            })
+            if(duplicate)
+            newCards.push({
+                cardId: Date.now(),
+                cardCVV: action.payload.cardCVV,
+                cardNumber: action.payload.cardNumber,
+                cardMonth: action.payload.cardMonth,
+                cardYear: action.payload.cardYear
+            })
+            console.log(newCards)
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/store/reducers/cardReducers.ts]: Delete it

File: src/components/card/CustomSelect.tsx [hunk 15]
```diff
@@ -0,0 +1,38 @@
+import React from 'react'
+import Select, {StylesConfig} from 'react-select'
+import {ISelectInput} from "../../types/cardTypes";
+import {PaymentsCardExpirationSelect} from './CardStyledComponents'
+
+
+function CustomSelect({handleChange, options,errors,onBlur,id,name,touched,placeholder, defaultValue}:ISelectInput) {
+     const customStyles: StylesConfig<any> = {
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/CustomSelect.tsx:8]: Refactor to Styled, but now onChange handler have any type, and i cant find what type is it

File: src/components/card/PaymentForm.tsx [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+        errors,
+        touched,
+        setFieldValue
+    }:FormikProps<IFormValues> = useFormik({
+        initialValues:initialValues,
+
+        onSubmit:(values) => {
+            console.log(values)
+    },
+        validationSchema:yup.object({
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/PaymentForm.tsx:43]: Moved schema to separate file

File: src/components/card/PaymentForm.tsx [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+            cardNumber:yup.string()
+                .typeError('Card number must contains only numbers!')
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/PaymentForm.tsx:45]: Even if we not show it to users? i used only red color, if error has occurred

File: src/components/card/PaymentForm.tsx [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+                                 label={<FormattedMessage
+                                     id={"app.card.body.creditCardCvv.main"}
+                                     defaultMessage={'CVV / CVC'}
+                                 />}
+                                 width={'auto'}
+                    />
+                    <PaymentsInputCVVText>
+                        <FormattedMessage
+                            id={"app.card.body.creditCardCvv.sub"}
+                            defaultMessage={'3 or 4 digits code'}
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/PaymentForm.tsx:142]: Default message? But it used only when id cannot be found

File: src/store/actions/translation.ts (hunk 2 above)
Comment (EgorIlinskiy) [reply src/store/actions/translation.ts:3]: Nope. Renamed it

File: src/types/cardTypes.ts (hunk 3 above)
Comment (EgorIlinskiy) [reply src/types/cardTypes.ts:16]: HandleChange we can. But for options and onblur even the official documentation writes unknown type.

File: src/components/card/CustomSelect.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/CustomSelect.tsx:8]: > Refactor to Styled, but now onChange handler have any type, and i cant find what type is it

Oh no... custom styles need to be in props. If we put it at styled, styles do not apply

File: src/components/translation/index.tsx [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+import {SetLanguagesContainer} from "./TranslationStyledComponents";
+
+const Translation:FC = (props)=>{
+
+    const local = navigator.language;
+    let language = useTypedSelector(selectLanguage);
+    const {SetEnglishLang, SetRussianLang} = useActions()
+
+    let text = English;
+   if(language === 'English') text = English
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/translation/index.tsx:19]: Refactor languages types to enum

File: src/components/translation/index.tsx (hunk 1 above)
Comment (EgorIlinskiy) [reply src/components/translation/index.tsx:20]: If we rewrite it to switch cases, will it help us? 
Or we need some another way

File: src/types/languageTypes.ts (hunk 4 above)
Comment (EgorIlinskiy) [reply src/types/languageTypes.ts:4]: Got it.

File: src/components/card/SavedCards.tsx [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+         return  <div key={card.cardId}
+                      className={'payments-card-saved-container'}
+                      onClick={()=>{
+                          SetPreselectedCard(card.cardId);
+                      }}
+
+         >
+                     <div>{card.cardNumber}</div>
+                    <div className={'payments-card-saved-row'}>
+                        <VisaLogo/>
```
[the remark this answers is not in the document]
Comment (EgorIlinskiy) [reply src/components/card/SavedCards.tsx]: We have that mechanism. Added to form visa and mastercard logo.

**Review by IlyaRucavitcyn — APPROVED** [review]

**Pull request merged**